src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -10,18 +10,41 @@
 using System.Runtime.Serialization;
 using Microsoft.AspNet.OData.Builder;
 using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Extensions;
 using Microsoft.OData;
 using Microsoft.OData.Edm;
+using System.Linq;
 
 namespace Microsoft.AspNet.OData.Formatter.Deserialization
 {
     internal static class DeserializationHelpers
     {
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
-        { 
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
-                       
+        {
+            IEdmProperty edmProperty;
+
+            // Extract edmProperty using a Case-Sensitive match
+            edmProperty = resourceType.FindProperty(property.Name);
+
+            if (readContext != null && readContext.EnableCaseInsensitiveModelBinding && edmProperty == null)
+            {
+                // if edmProperty is null, we try a Case-Insensitive match
```

Comment:
how can we distinguish case-insensitive property from the dynamic property? 

For example: 

Customer has property name "Name",  it can accept "name" as dynamic property. Now, can we?  

Can we involve @Michael Pizzo

Reply:
This has been take care of when I fixed @mikepizzo 's comment https://github.com/OData/WebApi/pull/2249#discussion_r489146141

1. We first do a case-sensitive match. This ensures that properties differentiated by case are matched correctly.
2. If a case-sensitive match fails and we have enabled case-insensitive binding, we try a case-insensitive match.
3. If we have multiple case-insensitive matches and no case sensitive matches, we throw an exception

Reply:
Just to be clear, if it doesn't match case-sensitive or case-insensitive, then it would be treated as a dynamic property.

Sam is right, though, that if the customer has a property named "Name", previously they could have specified a value "name" and it would have been interpreted as a dynamic property. Now we'll do the case-insensitive match and match it as the defined property "Name".

It would be poor design, and probably an accidental bug, to have dynamic properties that varied from defined property only by case, but having a way to disable case insensitive would address that very edge case.

Reply:
#Resolved